src/common/hankel.jl

```diff
@@ -11,5 +11,30 @@ momentum / spherical bessel order `l`, and the Hankel coordinate `p`.
 """
 function hankel(r::AbstractVector, r2_f::AbstractVector, l::Integer, p::T)::T where {T<:Real}
     integrand = r2_f .* sphericalbesselj_fast.(l, p .* r)
-    return 4T(π) * simpson(r, integrand)
+    4T(π) * simpson(r, integrand)
+end
+
+
+# This custom rule makes sure we can use loop vectorisation in the innermost (simpson) loop.
+function hankel(r::AbstractVector, r2_f::AbstractVector, l::Integer, p::TT) where {TT <: ForwardDiff.Dual}
+    # Note that
+    #   d H[f] / dp = 4\pi \int_0^∞ r^2 f(r) j_l'(p⋅r)⋅r dr
+    # and that
+    #   j_l'(x) = l / x * j_l(x) - j_{l+1}(x)
+
+    T  = ForwardDiff.valtype(TT)
+    pv = ForwardDiff.value(p)
+
+    jl = sphericalbesselj_fast.(l, pv .* r)
+    integrand_value = r2_f .* jl
+    value = 4T(π) * simpson(r, integrand_value)
+
+    if iszero(pv)
+        return TT(value, zero(T) * ForwardDiff.partials(p))
+    end
+
+    derivative_value = r2_f .* (l .* jl ./ pv
+                                - r .* sphericalbesselj_fast.(l+1, pv .* r))
+    derivative = 4T(π) * simpson(r, derivative_value)
+    TT(value, derivative * ForwardDiff.partials(p))
```

Comment:
I don't understand why you need this rule. Why not overload `simpson` and/or `sphericalbesselj_fast` individually? Less code, more functionality.

Reply:
Well, it does not fully address the issue. I tried a few things and that gets pretty messy as we have a couple of integration routines and I do want to avoid allocations. I'll try if by inlining the data generation into the simpson rule maybe it's not necessary to do this.

Reply:
Actually it turns out what hurts the most here is the allocations. Since it's either avoiding allocations or LoopVectorisation, I dropped LoopVectorisation (future support is on shaky grounds right now anyway).

I'm gonna leave the hankel workaround because it removes exactly the bottlenecks in a hand-optimised way without using `DiffResults` or some additional boilerplate. If you have a better idea, let me know. Note that just doing a custom rule for integration and bessels and passing the duals through does not cut it here (you need to pass 6 partials instead of just 1 partial to the bessel).